src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
- I like the implementation :)  I just prefer `CASOrdering` to `CasOrdering`, as CAS is already an acronym.

- for the name of the function, I suggest:
    + let the function return `Ok(previous)` when `previous` is the previous value.
    + let's name it `compare_exchange`.

  rationale: this proposal provides the most general interface for CAS-like atomic updates.  In the case that we only need a degenerated interface, the compiler should optimize out the costs incurred by the additional functionality.  In that case, `compare_exchange` seems the natural name for it.

  More thoughts: personally I think that returning `Ok(())` or `Ok(previous)` when successful, is not an important problem.  Information-wise, these two are exactly the same.  So I also think this function is actually a swap, since if the swap was successful, the caller realizes that `old` is actually the old value.  It seems I am not satisfied with `compare_and_set` because it usually means something weaker than `compare_and_swap`: returning only boolean values (successful or not).

  I think I tried to be as clear as possible on this issue.  I trust @stjepang 's best judgment., so please discuss with other members and decide on it.

Thank you for your efforts, and sorry for late responses!